nemo/core/callbacks.py

```diff
@@ -35,6 +54,463 @@
+            tensors_to_log (List of str or NmTensor): A list of either tensor names or NmTensors which will be logged
+                every step_freq steps.
+                Defaults to ["loss"] which only prints the loss.
+            wandb_name(str): wandb experiment name.
+                Defaults to None
+            wandb_project(str): wandb project name.
+                Defaults to None
+            args: argparse flags which will be logged as hyperparameters.
+                Defaults to None.
+            log_epoch (bool): Whether to log epoch and epoch training time to tensorboard.
+                Defaults to True.
+        """
+        if not _WANDB_AVAILABLE:
+            logging.error("Could not import wandb. Did you install it (pip install --upgrade wandb)?")
+        self._step_freq = step_freq
+        self._tensors_to_log = tensors_to_log
+        self._name = wandb_name
+        self._project = wandb_project
+        self._args = args
+        self._last_epoch_start = None
+        self._log_epoch = log_epoch
+
+    def on_train_start(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            if _WANDB_AVAILABLE and wandb.run is None:
+                wandb.init(name=self._name, project=self._project)
+                if self._args is not None:
+                    wandb.config.update(self._args)
+            elif _WANDB_AVAILABLE and wandb.run is not None:
+                logging.info("Re-using wandb session")
+            else:
+                logging.error("Could not import wandb. Did you install it (pip install --upgrade wandb)?")
+                logging.info("Will not log data to weights and biases.")
+                self._step_freq = -1
+
+    def on_step_end(self, state):
+        # log training metrics
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            if state["step"] % self._step_freq == 0 and self._step_freq > 0:
+                tensors_logged = {t: state["tensors"].get_tensor(t).cpu() for t in self._tensors_to_log}
+                # Always log learning rate
+                tensors_logged['LR'] = state["learning_rate"]
+                self._wandb_log(tensors_logged)
+
+    def on_epoch_start(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            self._last_epoch_start = time.time()
+
+    def on_epoch_end(self, state):
+        if state["global_rank"] is None or state["global_rank"] == 0:
+            if self._log_epoch:
+                epoch_time = time.time() - self._last_epoch_start
+                self._wandb_log({"epoch": state["epoch"], "epoch_time": epoch_time})
+
+    def _wandb_log(self, tensors_logged):
+        if _WANDB_AVAILABLE:
+            wandb.log(tensors_logged, step=state["step"])
+
+
+class CheckpointCallback(NeMoCallback):
+    def __init__(
+        self,
+        folder: str,
+        load_from_folder: str = None,
```

Comment:
Why inside of callback that is supposed to save we also have an option to load...?

Should _loading_ modules/checkpoint be part of callback system at all?

Reply:
chekpointing callback can resume training correctly (e.g. including optimizer state)